packages/react-jss/src/styled.js

```diff
@@ -0,0 +1,47 @@
+// @flow
+/* eslint-disable react/prop-types, react/require-default-props */
+
+import React from 'react'
+import isPropValid from '@emotion/is-prop-valid'
+import type {StatelessFunctionalComponent, ComponentType} from 'react'
+import withStyles from './withStyles'
+import type {Options, Style, Classes} from './types'
+
+// Props we don't want to forward.
+const reservedProps = {
+  classes: true,
+  as: true,
+  theme: true
+}
+
+type StyledProps = {
+  className?: string,
+  as?: string,
+  classes?: Classes
+}
+
+export default <Props: {}>(
+  type: string | StatelessFunctionalComponent<Props> | ComponentType<Props>
+) => {
+  const isTag = typeof type === 'string'
+
+  return <Theme: {}>(style: Style<Theme>, options?: Options<Theme>) => {
+    const Styled = (props: StyledProps) => {
+      const {classes, as, className} = props
+      const childProps: Props = ({}: any)
+      for (const prop in props) {
+        if (prop in reservedProps) continue
+        // We don't want to pass non-dom props to the DOM,
+        // but we still wat to forward them to a uses component.
+        if (isTag && !isPropValid(prop)) continue
+        childProps[prop] = props[prop]
+      }
+      // $FlowFixMe flow seems to not know that `classes` will be provided by the HOC, not by element creator.
+      childProps.className = className ? `${className} ${classes.css}` : classes.css
+      return React.createElement(as || type, childProps)
+    }
+
+    // $FlowFixMe flow seems to not know that `classes` will be provided by the HOC, not by element creator.
+    return withStyles({css: style}, {...options, injectTheme: true})(Styled)
```

Comment:
This implementation is good for now, but I would like to optimize it, later on, to not render two extra react components for example.

I'm not sure if we should use the hooks implementation in here yet, as this would make the `styled` API only usable with `react@16.8` currently, which in my opinion is a little bit early.

Reply:
I actually think performance is critical and this amount of components for each styled one is really bad, that's why I think hooks api is the only way forward with this API. For those who use older react versions they still can use the old styled implementation.